### PR TITLE
Factory for renderers

### PR DESCRIPTION
**Pull request by prytoegrian**
In order to contribute (add rules, correct bugs, …) in the future, I need to be more comfortable with your code, methods and practices. So I studied `Command` and `CommandLineOptions` and I decided to make some improvements to follow SRP in them. I also noted these classes might be agnostics of all renderers because at these steps, « renderable » notion isn't useful ; thus I created a `RendererFactory` to get a single entry point and display a common facade for consumers.

## Factory
I follow « Abstract Factory » design pattern in `RendererFactory`. I would also like to completely forbid constructions of `*Renderer` classes outside of `RendererFactory` entry point but I'm not sure it will stick to your coding convention so let me know your point of view about that.
I didn't test `RendererFactory` yet, I need to know if my way corresponds to yours methods and designs before.

  
I also enhance contract on `*Renderer` classes by making them immutables (*a render **need** a writer to work*).

PS : forgive my english, I'm rusted :yum:

### CODE REVIEW

**Comment by ravage84** [comment]
Hi @Prytoegrian 

Thanks for your effort.

I would need to discuss this with @manuelpichler .
But before we have to make sure, that you changes __need__ to be backward compatible.
If that's not the case, we can only release it by bumping the major version of PHPMD to 3.0.0.
If that's the case, your changes will not get released any time soon, as we do not have concrete plans for such a release.

**Comment by Prytoegrian** [comment]
Sure, take your time.

As far as I know, there is no compatibility break but perhaps your eye is sharper than mine, particularly on `*Renderer` public contract.

File: src/main/php/PHPMD/RendererFactory.php [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+     * <ul>
+     *   <li>xml</li>
+     *   <li>html</li>
+     *   <li>text</li>
+     * </ul>
+     *
+     * @param AbstractWriter $writer Asociated output writer instance
+     * @param string $reportFormat
+     * @return \PHPMD\AbstractRenderer
+     * @throws \InvalidArgumentException When the specified renderer does not exist.
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: It does not throw anything.

File: src/main/php/PHPMD/RendererFactory.php [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+     *
+     * Valid renderers are:
+     * <ul>
+     *   <li>xml</li>
+     *   <li>html</li>
+     *   <li>text</li>
+     * </ul>
+     *
+     * @param AbstractWriter $writer Asociated output writer instance
+     * @param string $reportFormat
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Display an example.

File: src/main/php/PHPMD/RendererFactory.php [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            case 'html':
+                return new HTMLRenderer($writer);
+            case 'text':
+                return new TextRenderer($writer);
+            default:
+                return static::createCustomRenderer($writer, $reportFormat);
+        }
+    }
+
+    /**
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Short description missing.

File: src/main/php/PHPMD/RendererFactory.php [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+     * @param AbstractWriter $writer Asociated output writer instance
+     * @param string $reportFormat
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Display an example.

File: src/main/php/PHPMD/RendererFactory.php [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+     * @return \PHPMD\AbstractRenderer
+     * @throws \InvalidArgumentException
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Description missing.

File: src/main/php/PHPMD/RendererFactory.php [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+     */
+    private static function createCustomRenderer(AbstractWriter $writer, $reportFormat)
+    {
+        if ('' === $reportFormat) {
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Do not use Yoda format.

File: src/main/php/PHPMD/RendererFactory.php [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+            throw new \InvalidArgumentException(
+                'Can\'t create report with empty format.'
+            );
+        }
+
+        if (class_exists($reportFormat)) {
+            return new $reportFormat($writer);
+        }
+
+        // Try to load a custom renderer
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Extract the follwoing code in a protected method.

File: src/main/php/PHPMD/RendererFactory.php [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        $fileName = strtr($reportFormat, '_\\', '//') . '.php';
+
+        $fileHandle = @fopen($fileName, 'r', true);
+        if (is_resource($fileHandle) === false) {
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Why don't you use ``file_exists()``?

File: src/main/php/PHPMD/RendererFactory.php [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        }
+
+        if (class_exists($reportFormat)) {
+            return new $reportFormat($writer);
+        }
+
+        // Try to load a custom renderer
+        $fileName = strtr($reportFormat, '_\\', '//') . '.php';
+
+        $fileHandle = @fopen($fileName, 'r', true);
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Using error suppression is a mess 😼

File: src/main/php/PHPMD/RendererFactory.php [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        if (is_resource($fileHandle) === false) {
+            throw new \InvalidArgumentException(
+                sprintf(
+                    'Can\'t find the custom report class: %s',
+                    $reportFormat
+                )
+            );
+        }
+        @fclose($fileHandle);
```
Comment (ravage84) [inline src/main/php/PHPMD/RendererFactory.php]: Using error suppression is a mess 😼

File: src/main/php/PHPMD/TextUI/Command.php [hunk 11]
```diff
@@ -156,6 +140,27 @@ private function getVersion()
     }
 
     /**
+     * Returns renderers asked by user
```
Comment (ravage84) [inline src/main/php/PHPMD/TextUI/Command.php]: Can you use a different word than "asked"? May be "requested"?

File: src/main/php/PHPMD/TextUI/Command.php [hunk 12]
```diff
@@ -156,6 +140,27 @@ private function getVersion()
...
+     * @return \PHPMD\AbstractRenderer[]
+     */
+    private function getRenderers(CommandLineOptions $opts)
+    {
+        $stream = $opts->getReportFile() ? fopen($opts->getReportFile(), 'wb') : STDOUT;
+        $renderers[] = $opts->createRenderer(new StreamWriter($stream));
+
+        foreach ($opts->getReportFiles() as $reportFormat => $reportFile) {
+            $renderers[] = $opts->createRenderer(
+                new StreamWriter(fopen($reportFile, 'wb')),
```
Comment (ravage84) [inline src/main/php/PHPMD/TextUI/Command.php]: What do you think about catching exceptions here?

**Comment by Prytoegrian** [comment]
> I'm still a bit concerned about breaking compatibility.
(*and some comments*)

Owh, indeed, I don't have the good mindset about compatibility break yet, *phpmd is a tool that can be used or extended*.
... I don't give up, I'll find a solution :rooster:

**Comment by ravage84** [comment]
@Prytoegrian  no problem.

Actually I'm not sure if it's really a real issue.
But if you find a solution that is BC, even better! 😸

**Comment by Prytoegrian** [comment]
Ok, I have a more precise vision on how to keep compatibility. But, I face a failed test and I'm not very confortable with your namespace convention so I'm lost.

```sh
1) PHPMD\TextUI\CommandLineOptionsTest::testCreateRenderer with data set #3 ('PHPMD_Test_Renderer_PEARRenderer', 'PHPMD_Test_Renderer_PEARRenderer')
InvalidArgumentException: Can't find the custom report class: PHPMD_Test_Renderer_PEARRenderer

.../phpmd/src/main/php/PHPMD/TextUI/CommandLineOptions.php:367
.../phpmd/src/test/php/PHPMD/TextUI/CommandLineOptionsTest.php:365

2) PHPMD\TextUI\CommandLineOptionsTest::testCreateRenderer with data set #4 ('PHPMD\Test\Renderer\NamespaceRenderer', 'PHPMD\Test\Renderer\NamespaceRenderer')
InvalidArgumentException: Can't find the custom report class: PHPMD\Test\Renderer\NamespaceRenderer

.../phpmd/src/main/php/PHPMD/TextUI/CommandLineOptions.php:367
.../phpmd/src/test/php/PHPMD/TextUI/CommandLineOptionsTest.php:365

3) PHPMD\TextUI\CommandLineOptionsTest::testCreateRenderer with data set #5 ('PHPMD\Test\Renderer\NamespaceRenderer', 'PHPMD\Test\Renderer\NamespaceRenderer')
InvalidArgumentException: Can't find the custom report class: PHPMD\Test\Renderer\NamespaceRenderer

.../phpmd/src/main/php/PHPMD/TextUI/CommandLineOptions.php:367
.../phpmd/src/test/php/PHPMD/TextUI/CommandLineOptionsTest.php:365

```

I would need help to make it pass :-/

**Comment by ravage84** [comment]
@Prytoegrian sorry, didn't have and don't have time to help you on this. I keep this PR open.

**Comment by ravage84** [comment]
Closing this after about 3 years being stale. I still like the idea.
@prytoegrian if you want you can open a new PR based on our current code which was refactored to some extend since back then.

**Pull request closed without merging**